code/modules/atmospherics/machinery/datum_pipeline.dm

```diff
@@ -210,40 +212,67 @@
 	var/list/datum/gas_mixture/GL = list()
 	var/list/datum/pipeline/PL = list()
 	PL += src
-
 	for(var/i = 1; i <= PL.len; i++) //can't do a for-each here because we may add to the list within the loop
 		var/datum/pipeline/P = PL[i]
 		if(!P)
 			continue
-		GL += P.return_air()
+		if(i == 1)
+			GL += P.other_airs
+		else
+			GL += P.return_air()
 		for(var/obj/machinery/atmospherics/components/binary/valve/V in P.other_atmosmch)
 			if(V.open)
 				PL |= V.parents[1]
 				PL |= V.parents[2]
 		for(var/obj/machinery/atmospherics/components/unary/portables_connector/C in P.other_atmosmch)
 			if(C.connected_device)
 				GL += C.portableConnectorReturnAir()
+	reconcile = FALSE
+	return air.update_pipeair(GL)
 
-	var/total_thermal_energy = 0
-	var/total_heat_capacity = 0
-	var/datum/gas_mixture/total_gas_mixture = new(0)
-
-	for(var/i in GL)
+/datum/gas_mixture/proc/update_pipeair(list/other_airs)
+	var/total_volume = volume
+	for(var/i in other_airs)
 		var/datum/gas_mixture/G = i
-		total_gas_mixture.volume += G.volume
-
-		total_gas_mixture.merge(G)
-
-		total_thermal_energy += THERMAL_ENERGY(G)
-		total_heat_capacity += G.heat_capacity()
-
-	total_gas_mixture.temperature = total_heat_capacity ? total_thermal_energy/total_heat_capacity : 0
-
-	if(total_gas_mixture.volume > 0)
-		//Update individual gas_mixtures by volume ratio
-		for(var/i in GL)
+		total_volume += G.volume
+		if(G.gases.len) //Average 0.6-0.8 false results for every proc call
+			merge(G)
+	var/list/cached_gases = gases
+	. = NO_REACTION
+	if(cached_gases.len) //Average 0.15 false results for every proc call
+		for(var/i in other_airs)	//Update individual gas_mixtures by volume ratio
 			var/datum/gas_mixture/G = i
-			G.copy_from(total_gas_mixture)
-			var/list/G_gases = G.gases
-			for(var/id in G_gases)
-				G_gases[id][MOLES] *= G.volume/total_gas_mixture.volume
+			var/ratio = G.volume/total_volume
+			G.temperature = temperature
+			for(var/id in cached_gases)
+				ASSERT_GAS(id,G)
+				G.gases[id][MOLES] = ratio * cached_gases[id][MOLES]
+		var/ratio = volume/total_volume
+		for(var/id in cached_gases)
+			cached_gases[id][MOLES] *= ratio
+		if(!length(cached_gases-GLOB.nonreactive_gases))
+			return
+		reaction_results = new
+		var/temp = temperature
+		var/ener = THERMAL_ENERGY(src)
+		reaction_loop:
+			for(var/r in SSair.gas_reactions)
+				var/datum/gas_reaction/reaction = r
+	
+				var/list/min_reqs = reaction.min_requirements.Copy()
+				if((min_reqs["TEMP"] && temp < min_reqs["TEMP"]) \
+				|| (min_reqs["ENER"] && ener < min_reqs["ENER"]))
+					continue
+				min_reqs -= "TEMP"
+				min_reqs -= "ENER"
+	
+				for(var/id in min_reqs)
+					if(!cached_gases[id] || cached_gases[id][MOLES] < min_reqs[id])
+						continue reaction_loop
+				. |= reaction.react(src, null)
+				if (. & STOP_REACTIONS)
+					break
+		if(.)
+			garbage_collect()
+			if(temperature < TCMB) //just for safety
+				temperature = TCMB
```

Comment:
what the Fuck

Reply:
Yea added react() they serve the same purpose and run faster here with the checks already in place. I'll look at adding the define wrapper we talked about to clean it up (and bring more of this beautiful SPEED to other procs).